test/unit-tests/decorators.ts

```diff
@@ -2,173 +2,437 @@ import * as decoratorsLib from "../../decorators";
-			_.each(promises, (promise, index) => promise.then((val: string) => {
-				assert.deepEqual(val, expectedResults[index]);
-			}, (err: Error) => {
-				// We cannot compare promise.reason() with error directly as node-fibers modify the error.stack property, so deepEqual method fails.
-				assert.deepEqual(err.message, expectedResults[index].message);
-			}));
+			promisifiedResultFunction({}, propertyName, {});
+			new Promise((onFulfilled: Function, onRejected: Function) => {
+				let promises: Promise<string>[] = $injector.publicApi.__modules__[moduleName][propertyName]();
+				_.each(promises, (promise, index) => promise.then((val: string) => {
+					assert.deepEqual(val, expectedResults[index]);
+					if (index + 1 === expectedResults.length) {
+						onFulfilled();
+					}
+				}, (err: Error) => {
+					// We cannot compare promise.reason() with error directly as node-fibers modify the error.stack property, so deepEqual method fails.
+					assert.deepEqual(err.message, expectedResults[index].message);
+					if (index + 1 === expectedResults.length) {
+						onFulfilled();
+					}
+				}));
+			}).then(done).catch(done);
+		});
+
+		describe("postAction", () => {
+			let isPostActionExecuted = false;
+			let isActionExecuted = false;
+			let expectedResults: any;
+
+			let postAction = () => {
+				assert.isTrue(isActionExecuted, "Post Action MUST be executed AFTER all actions are executed.");
+				isPostActionExecuted = true;
+			};
+
+			let getPromisesWithPostAction = (): any => {
+				let promisifiedResultFunction: any = decoratorsLib.exportedPromise(moduleName, postAction);
+				// Call this line in order to generate publicApi and get the real Promise
+				promisifiedResultFunction({}, propertyName, {});
+				return $injector.publicApi.__modules__[moduleName][propertyName]();
+			};
+
+			let assertResults = (result: any): void => {
+				assert.deepEqual(result, expectedResults);
+				assert.isTrue(isPostActionExecuted, "Post action must be executed after all promises are resolved.");
+				assert.isTrue(isActionExecuted, "All actions must be executed after the promise is resolved.");
+			};
+
+			beforeEach(() => {
+				isPostActionExecuted = false;
+				isActionExecuted = false;
+			});
+
+			it("executes postAction after all promises are resolved (function returning IFuture<T>)", (done) => {
+				expectedResults = "result";
+
+				$injector.register(moduleName, {
+					propertyName: () => {
+						return (() => {
+							assert.isFalse(isPostActionExecuted, "Post action MUST NOT be called before all actions are executed.");
+							isActionExecuted = true;
+							return expectedResults;
+						}).future<any>()();
+					}
+				});
+
+				let promise: any = getPromisesWithPostAction();
+				promise.then(assertResults)
+					.then(done)
+					.catch(done);
+			});
+
+			it("executes postAction after all promises are resolved (function returning IFuture<T>[])", (done) => {
+				let calledActionsCount = 0;
```

Comment:
You can almost definitely use the index provided by `_.map` instead of this custom variable